java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowUtils.java

```diff
@@ -221,16 +226,106 @@ private static Consumer createConsumer(
     return consumer;
   }
 
+
+  private static Consumer createSkippableConsumer(Schema schema) {
+
+    SkipFunction skipFunction;
+    Type type = schema.getType();
+
+    switch (type) {
+      case UNION:
+        int size = schema.getTypes().size();
+        long nullCount = schema.getTypes().stream().filter(s -> s.getType() == Type.NULL).count();
+        if (size == 1) {
+          return createSkippableConsumer(schema.getTypes().get(0));
+          // nullable primitive type
+        } else if (size == 2 && nullCount == 1) {
+          Schema nullSchema = schema.getTypes().stream().filter(s -> s.getType() == Type.NULL).findFirst().get();
+          int nullIndex = schema.getTypes().indexOf(nullSchema);
+          Schema subSchema = schema.getTypes().stream().filter(s -> s.getType() != Type.NULL).findFirst().get();
+          Preconditions.checkNotNull(subSchema, "schema should not be null.");
+          final Consumer consumer = createSkippableConsumer(subSchema);
+          return new NullableTypeConsumer(consumer, nullIndex);
+          // real union type
+        } else {
+
+          List<Consumer> delegates = schema.getTypes().stream().map(s ->
+              createSkippableConsumer(s)).collect(Collectors.toList());
+          skipFunction =  decoder -> delegates.get(decoder.readInt()).consume(decoder);
+
+          break;
+        }
+      case ARRAY:
+        skipFunction = decoder -> decoder.skipArray();
```

Comment:
I think this needs be in a loop.  Based on the java-doc:
for(long i = in.skipArray(); i != 0; i = i.skipArray()) {
     for (long j = 0; j < i; j++) {
       read and discard the next element of the array;
     }
   }